File: leetcode/rust-leetcode/src/735.asteroid-collision.rs
impl Solution {
    pub fn asteroid_collision(asteroids: Vec<i32>) -> Vec<i32> {
        let mut ans:Vec<i32> = Vec::new();
        for ast in asteroids {
            if ans.len() == 0 || ans[ans.len()-1] * ast > 0 {
                ans.push(ast);
                continue;
            }
            while ans.len() > 0 {
                let k = ans[ans.len() - 1];
                if k * ast > 0 || (ast > 0 && k < 0) {
                    ans.push(ast);
                    break;
                } else if ast < 0 && k > 0 {
                    if ast.abs() == k {
                        ans.pop();
                        break;
                    } else if ast.abs() > k {
                        ans.pop();
                        if ans.len() == 0 {
                            ans.push(ast);
                            break;
                        }
                    } else {
                        break;
                    }
                }
            }
        }
        ans
    }
}

File: leetcode/python/735.asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for a in asteroids:
            exploded = False
            while stack and stack[-1] > 0 and a < 0 and not exploded:
                if stack[-1] > abs(a):
                    exploded = True
                elif stack[-1] == abs(a):
                    exploded = True
                    stack.pop()
                else:
                    stack.pop()
            if not exploded:
                stack.append(a)
        return stack

